### PR TITLE
[FLINK-16008][table-planner][table-planner-blink] Add rules to transpose the join condition of Python Correlate node

### PR DESCRIPTION
**Pull request by HuangXingBo**
## What is the purpose of the change
 
*This pull request add rules to transpose the join condition of Python Correlate node*

## Brief change log

  - *Add rule CalcPythonCorrelateTransposeRule*
  - *Change match logic in SplitPythonConditionFromCorrelateRule*

## Verifying this change

This change added tests and can be verified as follows:

  - *CalcPythonCorrelateTransposeRuleTest*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

### CODE REVIEW

File: flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/rules/logical/CalcPythonCorrelateTransposeRule.java [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+	}
+
+	@Override
+	public boolean matches(RelOptRuleCall call) {
+		FlinkLogicalCorrelate correlate = call.rel(0);
+		FlinkLogicalCalc right = call.rel(2);
+		JoinRelType joinType = correlate.getJoinType();
+		FlinkLogicalCalc mergedCalc = StreamExecCorrelateRule.getMergedCalc(right);
+		FlinkLogicalTableFunctionScan scan = StreamExecCorrelateRule.getTableScan(mergedCalc);
+		return joinType == JoinRelType.INNER &&
```
Comment (hequn8128) [inline flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/rules/logical/CalcPythonCorrelateTransposeRule.java]: If we only support INNER, we need to also check the condition in `StreamExecPythonCorrelate` and throws a friendly error message to info users that currently correlate does not support conditions. 
What do you think?

File: flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/plan/rules/logical/CalcPythonCorrelateTransposeRuleTest.scala [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+import org.apache.flink.api.scala._
+import org.apache.flink.table.api.scala._
+import org.apache.flink.table.planner.plan.nodes.FlinkConventions
+import org.apache.flink.table.planner.plan.optimize.program._
+import org.apache.flink.table.planner.plan.rules.FlinkStreamRuleSets
+import org.apache.flink.table.planner.runtime.utils.JavaUserDefinedScalarFunctions.PythonScalarFunction
+import org.apache.flink.table.planner.utils.{MockPythonTableFunction, TableTestBase}
+import org.junit.{Before, Test}
+
+class CalcPythonCorrelateTransposeRuleTest extends TableTestBase{
```
Comment (hequn8128) [inline flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/plan/rules/logical/CalcPythonCorrelateTransposeRuleTest.scala]: Add a blank after the `TableTestBase `

File: flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/plan/CalcPythonCorrelateTransposeRuleTest.scala [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+package org.apache.flink.table.plan
+
+import org.apache.flink.api.scala._
+import org.apache.flink.table.api.scala._
+import org.apache.flink.table.runtime.utils.JavaUserDefinedScalarFunctions.PythonScalarFunction
+import org.apache.flink.table.utils.{MockPythonTableFunction, TableTestBase}
+import org.apache.flink.table.utils.TableTestUtil.{streamTableNode, term, unaryNode}
+import org.junit.Test
+
+class CalcPythonCorrelateTransposeRuleTest extends TableTestBase{
```
Comment (hequn8128) [inline flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/plan/CalcPythonCorrelateTransposeRuleTest.scala]: Add a blank after the `TableTestBase `

File: flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/rules/logical/CalcPythonCorrelateTransposeRule.java [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+			PythonUtil.isPythonCall(scan.getCall()) &&
```
Comment (hequn8128) [inline flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/rules/logical/CalcPythonCorrelateTransposeRule.java]: Check whether scan is defined? `scan.isDefined() &&`

**Comment by HuangXingBo** [comment]
Thanks a lot for @hequn8128 review, I have addressed the comments at the latest commit.

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/nodes/physical/stream/StreamExecPythonCorrelate.scala [hunk 5]
```diff
@@ -51,6 +52,10 @@ class StreamExecPythonCorrelate(
     joinType)
   with CommonPythonCorrelate {
 
+  if (condition.isDefined) {
+    throw new TableException("Currently Python correlate does not support conditions in left join")
```
Comment (hequn8128) [inline flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/nodes/physical/stream/StreamExecPythonCorrelate.scala]: Put a full comma at the end of the sentence.

**Comment by HuangXingBo** [comment]
Thanks a lot for @hequn8128 review, I have addressed the comments at the latest commit.

**Pull request merged**